### ISSUE TITLE
Design a new format of suppression dtd schema to support XPath queries

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#4364]
From my and @timurt point of view, XPath for queering AST nodes will be a good idea for the new suppression model, because:
1. XPath can be used for queering AST tree nodes in a simple and transparent way;
2. There are two ready to use open-source libraries which allow to work with XPath to query AST nodes easily (Jaxen and Saxon). Both libraries are used by PMD, so we can adopt their experience. No need to reinvent the wheel;
3. XPath is a good solution for xml-format as other tools such as PMD and Sonarqube already support it, and it will be easy for the programmer to write suppression rules for the Checkstyle’s violations if the format of the suppressions will be similar.
4. User who is already familiar with AST tokens we use in the checks, can have the base understanding of tree structure which will help him to write XPath expression. 

Based on the arguments, mentioned above, we decided that we need to use XPath in the new model.

**What should be done in the scope of this issue:**
1. The decision on whether to extend [the existing dtd schema](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/suppressions_1_1.dtd) or create a new one specially for the new suppression model should be made;
2. In case we decide that we need to extend the existing schema, the new format of the suppression element need to be proposed, otherwise the new format of dtd schema should be introduced.

My proposal:
1. We need to extend the existing dtd schema to avoid breaking of backward compatibility. When we introduce the new suppression model nobody should notice the problem. The projects which use the old line-based approach should have an ability to continue suppressing Checkstyle violations by lines.
2. In his [GSoC 2017 proposal](https://docs.google.com/document/d/1nwQmXRY7FQoXB4icdnstnOOc8aloJSiLKDXfEWBGHbk/edit#heading=h.8hhjmh38bbo2) @timurt suggested adding a new attribute for the existing suppression element:

```xml
<suppress-xpath
  checks="FileLength"
  files="TokenTypes.java"
  query="xpath expression here”
/>
```

I think that this is a good idea, but from my point of view, "xpath" would be a better name for the attribute. Such approach will allow us to support two models in the same file. User will not need to rewrite the whole suppression file from scratch.

Other solution is to add a new attribute to the existing "suppress" element:

```xml
<suppression
   checks="FileLength"
   lines="1,2,3"
   files="TokenTypes.java"
   xpath="xpath expression here”
/>
```

In the second approach the influence on dtd schema will be minimal.

**Warning**
In this issue we do not discuss all details of the implementation. We just discuss the new format of the dtd schema which will be a user-friendly and will not break backward compatibility.

**Comment by MEZk** [comment]
@romani, @rnveach, @Vladlis, @sabaka, @timurt, please, join the discussion.

**Comment by rnveach** [comment]
Just my opinion,

> User who is already familiar with AST tokens we use in the checks, can have the base understanding of tree structure which will help him to write XPath expression.

These type of users will be rare outside our own development team. Some people probably have delved into writing their own checks, but I expect most don't.
Some of our AST tree doesn't make sense, even to us, and we have open issues to fix them.
Also, any changes to our grammar has the possibility to break xpath suppression especially if we move to Antlr4 for Java.

> extend the existing dtd schema or create a new one 
> Other solution is to add a new attribute to the existing "suppress" element

I like this solution the best.

> the new format of the suppression element 
> <suppression ... xpath="xpath expression here”

Again, I like this solution.

> Such approach will allow us to support two models in the same file.

I don't see it as 2 separate models, just that we are adding a new feature onto the existing model.
For example, I see no reason you can't use `xpath` and `files` together. I assume only file's AST will be part of the xpath and not external things like file/folder name.

**Comment by MEZk** [comment]
@rnveach 
Thanks for the answer.

> Also, any changes to our grammar has the possibility to break xpath suppression especially if we move to Antlr4 for Java.

The original GSoC task says that 

> We need to invent new Suppression model that is flexible to code changes. This will allow user to have legacy code and new code in the same file and not suffer from updating suppression configuration on any code changes. Unlike of suppression that is based on lines' numbers in file, new model should be based on [Abstract Syntax Tree (AST) ](http://en.wikipedia.org /wiki/Abstract_syntax_tree) structure.

Thus, the new model should be based on the AST structure. I could not find any better ways to refer AST structure, except for XPath.   We do not change token types names very often. Thus, from my point of view, we can rely on their names in the xpath expression. I thought about other solutions, for example, in the issue #2136 it is requested to add the new attribute "methods" to suppress methods with the specified names. This approach is not good as it will require to support many arguments to suppress violations for classes, methods, variables, and other language units. In the issue #2804 you suggest suppressing by message. However, it will require writing of regular expressions and sometimes the expression may become bigger and bigger, for example in [import control configuration](https://github.com/MEZk/checkstyle/blob/ddeb553834f10570089a033fa1ac1501b74ac358/config/import-control.xml#L140-L140).

> These type of users will be rare outside our own development team. Some people probably have delved into writing their own checks, but I expect most don't.

That is why it is required to create suppression file generator to ease suppression configuration creation for legacy code. I will increase the priority of this task.

**Comment by MEZk** [comment]
@romani 
Are you OK with this format:

```xml
<suppression
   checks="FileLength"
   lines="1,2,3"
   files="TokenTypes.java"
   xpath="xpath expression here”
/>
```

**Comment by rnveach** [comment]
>  We do not change token types names very often.

@MEZk I was referring to changing the AST structure, not names. As an example, we have 1 issue opened for variable declarations that are wrong which may be a common suppression area. https://github.com/checkstyle/checkstyle/issues/3151

I was not specifically saying we shouldn't do this xpath suppression with AST, I was just stating one problem that will arise with this and break compatibility between releases.

**Comment by timurt** [comment]
@MEZk @rnveach 
I just wanted to clarify, 

```
<suppression
   checks="FileLength"
   lines="1,2,3"
   files="TokenTypes.java"
   xpath="xpath expression here”
/>
```
Here `lines` and `xpath` attributes have not effect on each other, it is like two suppression models in one?

**Comment by MEZk** [comment]
@romani 
ping

**Comment by rnveach** [comment]
@MEZk 
if 
````
<suppression
   checks="FileLength"
   lines="1,2,3"
   files="TokenTypes.java"
   xpath="xpath expression here”
/>
````
is handled the same as 
````
<suppression
   xpath="xpath expression here”
/>
<suppression
   checks="FileLength"
   lines="1,2,3"
   files="TokenTypes.java"
/>
````
then the combined suppression seems misleading to me as I expected all attributes to be combined together, the same way `checks` and `lines` does now.
This is the way I understood the proposed model when we specified they would be all together.

If you are saying this is not the case, then I see no reason to keep them together, even though I would like that way better.
It hasn't been specified what we can query on for xpath. I assumed file name would not be possible in xpath since it is only Java tree. So how would I be able to specify a specific xpath for a specific file to avoid conflicts over many files?

**Comment by MEZk** [comment]
@rnveach 
> then the combined suppression seems misleading to me as I expected all attributes to be combined together, the same way files and lines does now.
This is the way I understood the proposed model when we specified they would be all together.

Lines and Xpath should never be combined. Xpath model is opposite to line-based. Maybe you are right and it is better to add a new element to the schema (suppression-xpath) to make it clear to the user that we do not support mixing of line-based suppressions with xpath. The schema should validate that suppressiom-xpath element does not have lines attribute.

> It hasn't been specified what are the limits of what we can query on for xpath. I assumed file name would not be possible in xpath since it is only based on the Java tree. So how would I be able to specify a specific xpath for a specific file to avoid conflicts over many files?

There can be 2 or more source files with the same class name, for example.  ```suppression-xpath``` without file name attribute will require to specify package name token to distinguish two classes with the same name. Is it better than just specifying the file name as the attribute? However, if we implement truly xpath suppressions, we should rely only on tree structure, so we need to avoid ```files``` attribute for ```xpath-suppression``` element.

**Comment by rnveach** [comment]
> So, lets use files attribute.
> Thus, lets use checks attribute 

I agree.

> Lines and Xpath should never be combined. 

While this is probably generally true, I don't see a downside to allowing it, but I don't know the full syntax we will support in XPath. `lines` allows ranges. Even, if XPath supports everything lines does, using lines attribute will still be faster as fail first before doing xpath parsing.

> They will be processed separately by different filters.

As discussed in chat, `CsvFilter` and the like are actually, what I will call for now, sub-filters. So you will be creating a new sub-filter and not a new filter (like `SuppressionFilter`) for the xpaths.

> it is better to add a new element to the schema (suppression-xpath)

You will still be using `SuppressElement` which has lines in it's declaration, so you still have to 'technically' support it no matter what. The only thing the new tag is for is to dis-allow putting lines and xpath together in suppression, which as I stated above, I don't see it as a bad thing.

**Comment by MEZk** [comment]
@timurt 
1. Please add the new element to the suppression dtd schema;
2. Bump schema version;
3. Add schema to catalog file.

**Comment by romani** [comment]
@timurt , @rnveach , @MEZk 
Sorry for delay.

1)
unfortunately dtd syntax does not allow "choice" for attributes - https://www.w3.org/TR/REC-xml/#attdecls
so we can restrict to use `line` or `xpath` by dtd. We cannot allow the both attributes to exists as they are on the same level of match - it will not be intuitive for user.
So new tag is ok to appear.

2)
as new schema is about to appear please make its version as experimental `suppressions_1_1_xpath_experimental.dtd`, register it in code, but do not update other code to use it. It should not be public till we ready to share Xpath functionality with community.

3)
>
```
<suppression-xpath
  checks="MethodLength"
  files="MyClass.java, FooBar.java"
  query="//ClassDef//MethodDef[foo]”
/>
```

I am agree of format. `files` has to be but as optional attribute, as it is a different level of match could be used for folder match, example `<suppress checks="IllegalCatch" files="[\\/]internal[\\/]\w+Util\.java"/>` from https://github.com/checkstyle/checkstyle/blob/master/config/suppressions.xml#L33
or `<suppress checks="WriteTag" files=".*[\\/]src[\\/](test|it)[\\/]"/>`

4)
> . I will increase the priority of this task.

@MEZk , no need. Lets make Xpath work first. If any changes happen to AST structure, user will need to relaunch suppression generator. We will think about how to ease users life when we come to this problem, we will have this problem ourself first :) (as we eat food that we produce).

**Comment by timurt** [comment]
@MEZk @romani 
should I include new dtd inside catalog.xml?

**Comment by romani** [comment]
@timurt , yes

**Comment by MEZk** [comment]
@romani 
> files has to be but as optional attribute, as it is a different level of match could be used for folder match

Yes, it will be the optional attribute. XPath will allow to refer to the package name.

**Comment by rnveach** [comment]
@MEZk if this issue is done, please close it.

**Comment by rnveach** [comment]
@romani Please confirm, is this just a miscellaneous change?

**Comment by romani** [comment]
yes, we do not want to make it public, so it is internal update , so "misc" label should be used to move it in releasenotes to "Notes" group
